identity/src/main/java/androidx/security/identity/GattClient.java

```diff
@@ -201,27 +244,107 @@ public void onMtuChanged(@NonNull BluetoothGatt gatt, int mtu, int status) {
+            if (!mSupportL2CAP) {
+                reportError(new Error("Unexpected read for L2CAP characteristic "
+                        + characteristic.getUuid() + ", L2CAP not supported"));
+                return;
+            }
+            byte[] psmValue = characteristic.getValue();
+            if (psmValue == null || psmValue.length == 0 || psmValue.length > 4) {
+                reportError(new Error("Invalid PSM value received on L2CAP characteristic"));
+                return;
+            }
+            byte[] psmSized = new byte[4];
+            if (psmValue.length < 4) {
+                // Add 00 on left if psm length is lower than 4
+                System.arraycopy(psmValue, 0, psmSized, 4 - psmValue.length, psmValue.length);
+            } else {
+                psmSized = psmValue;
+            }
+            int psm = ByteBuffer.wrap(psmSized).getInt();
+            if ((mLoggingFlags & Constants.LOGGING_FLAG_TRANSPORT_SPECIFIC) != 0) {
+                Log.d(TAG, "Received psmValue: " + Util.toHex(psmValue) + " psm: " + psm);
+            }
+            try {
+                mSocket = mGatt.getDevice().createInsecureL2capChannel(psm);
+                mSocket.connect();
+                if (mSocket.isConnected()) {
+                    if ((mLoggingFlags & Constants.LOGGING_FLAG_TRANSPORT_SPECIFIC) != 0) {
+                        Log.d(TAG, "Connected using L2CAP on PSM: " + psm);
+                    }
+                    reportPeerConnected();
+                    listenSocket();
+                } else {
+                    reportError(new Error("Unable to connect L2CAP socket"));
+                }
+            } catch (IOException e) {
+                Log.e(TAG, "Error connecting to socket L2CAP " + e.getMessage(), e);
+                reportError(new Error("Error connecting to socket L2CAP"));
+            }
+        } else {
             reportError(new Error("Unexpected onCharacteristicRead for characteristic "
-                + characteristic.getUuid() + ", expected " + mCharacteristicIdentUuid));
+                    + characteristic.getUuid() + ", expected " + mCharacteristicIdentUuid));
             return;
         }
-        byte[] identValue = characteristic.getValue();
+    }
+
+    private void listenSocket() {
         if ((mLoggingFlags & Constants.LOGGING_FLAG_TRANSPORT_SPECIFIC) != 0) {
-            Log.d(TAG, "Received identValue: " + Util.toHex(identValue));
-        }
-        // TODO: maybe comment out or change to warning since it's optional... several readers
-        //  send the wrong value (others send the right one though)
-        if (!Arrays.equals(identValue, mIdentValue)) {
-            reportError(new Error("Received ident does not match expected ident"));
-            return;
+            Log.i(TAG, "Start reading socket input");
+        }
+        // We can check better/right buffer size to use with L2CAP
+        byte[] mmBuffer = new byte[1024 * 16];
```

Comment:
Current implementation will have issue if data from mdl holder is in multiple of your buffer length , in that case it will not be able to identify the last packet 

Current data length from mdl holder is 12550 , on using the same here it will get stuck in socket